File: src/cfg.fs
module Cfg

type SimpleInstr =
    | Label of string
    | ConditionalJump of string
    | UnconditionalJump of string
    | Return
    | Other

[<CustomEquality; CustomComparison>]
type NodeId =
    | Entry
    | Block of int
    | Exit

    override this.Equals(obj) =
        match obj with
        | :? NodeId as other ->
            match (this, other) with
            | Entry, Entry -> true
            | Block a, Block b -> a = b
            | Exit, Exit -> true
            | _ -> false
        | _ -> false

    override this.GetHashCode() =
        match this with
        | Entry -> 0
        | Block n -> hash (1, n)
        | Exit -> 2

    interface System.IComparable with
        member this.CompareTo(obj) =
            let tag =
                function
                | Entry -> 0
                | Block _ -> 1
                | Exit -> 2
            match obj with
            | :? NodeId as other ->
                match (this, other) with
                | Entry, Entry -> 0
                | Exit, Exit -> 0
                | Block a, Block b -> compare a b
                | _ -> compare (tag this) (tag other)
            | _ ->
                invalidArg "obj" "Cannot compare values of different types"

type BasicBlock<'v, 'instr> = {
    id: NodeId
    instructions: ('v * 'instr) list
    mutable preds: NodeId list
    mutable succs: NodeId list
    value: 'v
}

type ControlFlowGraph<'v, 'instr> = {
    (* store basic blocks in association list, indexed by block # *)
    BasicBlocks: (int * BasicBlock<'v, 'instr>) list
    mutable entry_succs: NodeId list
    mutable exit_preds: NodeId list
    debug_label: string
}

let get_succs nd_id cfg =
    match nd_id with
    | Entry -> cfg.entry_succs
    | Block n ->
        let nd = List.find (fun (k, _) -> k = n) cfg.BasicBlocks |> snd
        nd.succs
    | Exit -> []

let get_block_value blocknum cfg =
    let nd =
        List.find (fun (k, _) -> k = blocknum) cfg.BasicBlocks |> snd
    nd.value

let private update_successors f nd_id g =
    match nd_id with
    | Entry -> g.entry_succs <- f g.entry_succs
    | Block n ->
        let blk =
            List.find (fun (k, _) -> k = n) g.BasicBlocks |> snd
        blk.succs <- f blk.succs
    | Exit -> failwith "Internal error: malformed CFG"

let private update_predecessors f nd_id g =
    match nd_id with
    | Entry -> failwith "Internal error: malformed CFG"
    | Block n ->
        let blk =
            List.find (fun (k, _) -> k = n) g.BasicBlocks |> snd
        blk.preds <- f blk.preds
    | Exit -> g.exit_preds <- f g.exit_preds

let add_edge pred succ g =
    let add_id nd_id id_list =
        if List.contains nd_id id_list then id_list
        else nd_id :: id_list
    update_successors (add_id succ) pred g
    update_predecessors (add_id pred) succ g

let remove_edge pred succ g =
    let remove_id nd_id id_list =
        List.filter (fun i -> i <> nd_id) id_list
    update_successors (remove_id succ) pred g
    update_predecessors (remove_id pred) succ g

(* replace block with given block ID *)
let update_BasicBlock block_idx new_block g =
    let new_blocks =
        List.map
            (fun ((i, _) as blk) ->
                if i = block_idx then (i, new_block) else blk)
            g.BasicBlocks
    { g with BasicBlocks = new_blocks }

(* constructing the CFG *)
let private partition_into_BasicBlocks simplify instructions =
    let f (finished_blocks, current_block) i =
        match simplify i with
        | Label _ ->
            let finished_blocks' =
                if current_block = [] then finished_blocks
                else List.rev current_block :: finished_blocks
            (finished_blocks', [ i ])
        | ConditionalJump _ | UnconditionalJump _ | Return ->
            let finished = List.rev (i :: current_block)
            (finished :: finished_blocks, [])
        | Other -> (finished_blocks, i :: current_block)
    let finished, last = List.fold f ([], []) instructions
    let all_blocks =
        if last = [] then finished else List.rev last :: finished
    List.rev all_blocks

let private add_all_edges simplify g =
    (* build map from labels to the IDs of the blocks that they start with *)
    let label_map =
        List.fold
            (fun lbl_map (_, blk) ->
                match simplify (snd (List.head blk.instructions)) with
                | Label lbl -> Map.add lbl blk.id lbl_map
                | _ -> lbl_map)
            Map.empty g.BasicBlocks

    (* add outgoing edges from a single basic block *)
    let process_node (id_num, block) =
        let next_block =
            if id_num = fst (ListUtil.last g.BasicBlocks) then Exit
            else Block(id_num + 1)
        let _, last_instr = ListUtil.last block.instructions

        match simplify last_instr with
        | Return -> add_edge block.id Exit g
        | UnconditionalJump target ->
            let target_id = Map.find target label_map
            add_edge block.id target_id g
        | ConditionalJump target ->
            let target_id = Map.find target label_map
            add_edge block.id next_block g
            add_edge block.id target_id g
        | _ -> add_edge block.id next_block g

    add_edge Entry (Block 0) g
    List.iter process_node g.BasicBlocks

let instructions_to_cfg simplify debug_label instructions =
    let to_node idx instructions =
        let ann x = ((), x)
        (idx,
         { id = Block idx
           instructions = List.map ann instructions
           preds = []
           succs = []
           value = () })
    let cfg =
        { BasicBlocks =
              List.mapi to_node
                  (partition_into_BasicBlocks simplify instructions)
          entry_succs = []
          exit_preds = []
          debug_label = debug_label }

    add_all_edges simplify cfg
    cfg

(* converting back to instructions *)
let cfg_to_instructions g =
    let blk_to_instrs (_, { instructions = instructions }) =
        List.map snd instructions
    List.collect blk_to_instrs g.BasicBlocks

(* working with annotations *)
(* NOTE: Cannot use { x with ... } here because F# doesn't allow changing
   the type parameter in a record update expression (unlike OCaml).
   We construct new records explicitly to allow 'v to change. *)
let initialize_annotation cfg dummy_val =
    let initialize_instruction (_, i) = (dummy_val, i)
    let initialize_block (idx, b) =
        (idx,
         { id = b.id
           instructions = List.map initialize_instruction b.instructions
           preds = b.preds
           succs = b.succs
           value = dummy_val })
    { BasicBlocks = List.map initialize_block cfg.BasicBlocks
      entry_succs = cfg.entry_succs
      exit_preds = cfg.exit_preds
      debug_label = cfg.debug_label }

let strip_annotations cfg = initialize_annotation cfg ()

(* debugging *)
let print_graphviz (pp_instr: System.IO.TextWriter -> 'instr -> unit)
                   (pp_val: System.IO.TextWriter -> 'v -> unit)
                   (cfg: ControlFlowGraph<'v, 'instr>) =
    let filename =
        UniqueIds.makeLabel cfg.debug_label + ".dot"
    let path =
        if System.IO.Path.IsPathRooted(filename) then filename
        else System.IO.Path.Combine(System.IO.Directory.GetCurrentDirectory(), filename)
    use writer = new System.IO.StreamWriter(path)
    let pp_NodeId (out: System.IO.TextWriter) = function
        | Exit -> out.Write("exit")
        | Entry -> out.Write("entry")
        | Block n -> out.Write(sprintf "block%d" n)
    let pp_annotated_instruction (out: System.IO.TextWriter) (v, i) =
        out.Write("<tr>")
        out.Write("<td align=\"left\">")
        pp_instr out i
        out.Write("</td>")
        out.Write("<td align=\"left\">")
        pp_val out v
        out.Write("</td>")
        out.Write("</tr>")
        out.WriteLine()
    let pp_block_instructions (out: System.IO.TextWriter) blk =
        out.Write("<table>")
        out.Write("<tr><td colspan=\"2\"><b>")
        pp_NodeId out blk.id
        out.Write("</b></td></tr>")
        out.WriteLine()
        List.iter (pp_annotated_instruction out) blk.instructions
        out.Write("<tr><td colspan=\"2\">")
        pp_val out blk.value
        out.Write("</td></tr>")
        out.Write("</table>")
    let pp_block (out: System.IO.TextWriter) (lbl, b) =
        out.Write(sprintf "block%d[label=<" lbl)
        pp_block_instructions out b
        out.Write(">]")
    let pp_entry_edge (out: System.IO.TextWriter) lbl =
        out.Write("entry -> ")
        pp_NodeId out lbl
    let pp_edge i (out: System.IO.TextWriter) succ =
        out.Write(sprintf "block%d -> " i)
        pp_NodeId out succ
    let pp_edges (out: System.IO.TextWriter) ((lbl: int), (blk: BasicBlock<'a, _>)) =
        List.iter (pp_edge lbl out) blk.succs
    writer.WriteLine("digraph {")
    writer.WriteLine("  labeljust=l")
    writer.WriteLine("  node[shape=\"box\"]")
    writer.WriteLine("  entry[label=\"ENTRY\"]")
    writer.WriteLine("  exit[label=\"EXIT\"]")
    List.iter (fun b -> pp_block writer b; writer.WriteLine()) cfg.BasicBlocks
    List.iter (fun e -> pp_entry_edge writer e; writer.WriteLine()) cfg.entry_succs
    List.iter (fun b -> pp_edges writer b; writer.WriteLine()) cfg.BasicBlocks
    writer.WriteLine("}")
    writer.Flush()
    writer.Close()
    let cmd =
        sprintf "dot -Tpng %s -o %s" filename
            (System.IO.Path.ChangeExtension(filename, ".png"))
    let proc = System.Diagnostics.Process.Start("bash", "-c \"" + cmd + "\"")
    let finished = proc.WaitForExit(30000)
    if not finished then failwith ("graphviz fail: " + cmd)


module TackyCfg =
    let simplify = function
        | Tacky.Label l -> Label l
        | Tacky.Jump target -> UnconditionalJump target
        | Tacky.JumpIfZero(_, target) -> ConditionalJump target
        | Tacky.JumpIfNotZero(_, target) -> ConditionalJump target
        | Tacky.Return _ -> Return
        | _ -> Other

    let instructions_to_cfg debug_label instructions =
        instructions_to_cfg simplify debug_label instructions

    let cfg_to_instructions g = cfg_to_instructions g
    let get_succs nd_id cfg = get_succs nd_id cfg
    let get_block_value blocknum cfg = get_block_value blocknum cfg
    let add_edge pred succ g = add_edge pred succ g
    let remove_edge pred succ g = remove_edge pred succ g
    let update_BasicBlock idx blk g = update_BasicBlock idx blk g
    let initialize_annotation cfg v = initialize_annotation cfg v
    let strip_annotations cfg = strip_annotations cfg

module AsmCfg =
    let simplify = function
        | Assembly.Label l -> Label l
        | Assembly.Jmp target -> UnconditionalJump target
        | Assembly.JmpCC(_, target) -> ConditionalJump target
        | Assembly.Ret -> Return
        | _ -> Other

    let instructions_to_cfg debug_label instructions =
        instructions_to_cfg simplify debug_label instructions

    let cfg_to_instructions g = cfg_to_instructions g
    let get_succs nd_id cfg = get_succs nd_id cfg
    let get_block_value blocknum cfg = get_block_value blocknum cfg
    let add_edge pred succ g = add_edge pred succ g
    let remove_edge pred succ g = remove_edge pred succ g
    let update_BasicBlock idx blk g = update_BasicBlock idx blk g
    let initialize_annotation cfg v = initialize_annotation cfg v
    let strip_annotations cfg = strip_annotations cfg

File: src/Cfg.rs
//! Generic control-flow graphs, faithfully ported from the original `Cfg.fs`.
#![allow(non_snake_case)]

use std::collections::BTreeMap;
use std::fs::File;
use std::io::{self, Write};
use std::process::Command;
use std::sync::atomic::{AtomicUsize, Ordering};

/// The control-flow effect of an instruction.
#[derive(Clone, Debug, Eq, PartialEq)]
pub enum SimpleInstr {
    Label(String),
    ConditionalJump(String),
    UnconditionalJump(String),
    Return,
    Other,
}

/// An entry node, numbered basic block, or exit node.
///
/// Variant order deliberately matches the custom comparison in the F# type.
#[derive(Clone, Copy, Debug, Eq, Hash, Ord, PartialEq, PartialOrd)]
pub enum NodeId {
    Entry,
    Block(usize),
    Exit,
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct BasicBlock<V, I> {
    pub id: NodeId,
    pub instructions: Vec<(V, I)>,
    pub preds: Vec<NodeId>,
    pub succs: Vec<NodeId>,
    pub value: V,
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct ControlFlowGraph<V, I> {
    /// Basic blocks are kept in an association list indexed by block number.
    pub BasicBlocks: Vec<(usize, BasicBlock<V, I>)>,
    pub entrySuccs: Vec<NodeId>,
    pub exitPreds: Vec<NodeId>,
    pub debugLabel: String,
}

fn findBlock<V, I>(n: usize, blocks: &[(usize, BasicBlock<V, I>)]) -> &BasicBlock<V, I> {
    blocks
        .iter()
        .find(|(key, _)| *key == n)
        .map(|(_, block)| block)
        .expect("Internal error: block not found in CFG")
}

pub fn getSuccs<V, I>(ndId: NodeId, cfg: &ControlFlowGraph<V, I>) -> &[NodeId] {
    match ndId {
        NodeId::Entry => &cfg.entrySuccs,
        NodeId::Block(n) => &findBlock(n, &cfg.BasicBlocks).succs,
        NodeId::Exit => &[],
    }
}

pub fn getBlockValue<V, I>(blocknum: usize, cfg: &ControlFlowGraph<V, I>) -> &V {
    &findBlock(blocknum, &cfg.BasicBlocks).value
}

fn updateBlock<V, I, F>(f: F, blockNum: usize, g: &mut ControlFlowGraph<V, I>)
where
    F: FnOnce(&mut BasicBlock<V, I>),
{
    if let Some((_, block)) = g.BasicBlocks.iter_mut().find(|(i, _)| *i == blockNum) {
        f(block);
    }
}

fn addId(ndId: NodeId, ids: &mut Vec<NodeId>) {
    if !ids.contains(&ndId) {
        // F#'s list cons makes newly discovered edges appear first.
        ids.insert(0, ndId);
    }
}

fn removeId(ndId: NodeId, ids: &mut Vec<NodeId>) {
    ids.retain(|id| *id != ndId);
}

fn updateSuccessors<V, I, F>(f: F, ndId: NodeId, g: &mut ControlFlowGraph<V, I>)
where
    F: FnOnce(&mut Vec<NodeId>),
{
    match ndId {
        NodeId::Entry => f(&mut g.entrySuccs),
        NodeId::Block(n) => updateBlock(|block| f(&mut block.succs), n, g),
        NodeId::Exit => panic!("Internal error: malformed CFG"),
    }
}

fn updatePredecessors<V, I, F>(f: F, ndId: NodeId, g: &mut ControlFlowGraph<V, I>)
where
    F: FnOnce(&mut Vec<NodeId>),
{
    match ndId {
        NodeId::Entry => panic!("Internal error: malformed CFG"),
        NodeId::Block(n) => updateBlock(|block| f(&mut block.preds), n, g),
        NodeId::Exit => f(&mut g.exitPreds),
    }
}

/// Add an edge, without duplicating an existing predecessor or successor.
pub fn addEdge<V, I>(pred: NodeId, succ: NodeId, g: &mut ControlFlowGraph<V, I>) {
    updateSuccessors(|ids| addId(succ, ids), pred, g);
    updatePredecessors(|ids| addId(pred, ids), succ, g);
}

pub fn removeEdge<V, I>(pred: NodeId, succ: NodeId, g: &mut ControlFlowGraph<V, I>) {
    updateSuccessors(|ids| removeId(succ, ids), pred, g);
    updatePredecessors(|ids| removeId(pred, ids), succ, g);
}

/// Replace the block with the specified association-list key.
pub fn updateBasicBlock<V, I>(
    blockIdx: usize,
    newBlock: BasicBlock<V, I>,
    g: &mut ControlFlowGraph<V, I>,
) {
    if let Some(entry) = g.BasicBlocks.iter_mut().find(|(i, _)| *i == blockIdx) {
        entry.1 = newBlock;
    }
}

fn partitionIntoBasicBlocks<I, F>(simplify: &F, instructions: Vec<I>) -> Vec<Vec<I>>
where
    F: Fn(&I) -> SimpleInstr,
{
    let mut finished = Vec::new();
    let mut current = Vec::new();
    for instruction in instructions {
        match simplify(&instruction) {
            SimpleInstr::Label(_) => {
                if !current.is_empty() {
                    finished.push(current);
                    current = Vec::new();
                }
                current.push(instruction);
            }
            SimpleInstr::ConditionalJump(_)
            | SimpleInstr::UnconditionalJump(_)
            | SimpleInstr::Return => {
                current.push(instruction);
                finished.push(current);
                current = Vec::new();
            }
            SimpleInstr::Other => current.push(instruction),
        }
    }
    if !current.is_empty() {
        finished.push(current);
    }
    finished
}

fn addAllEdges<I, F>(simplify: &F, g: &mut ControlFlowGraph<(), I>)
where
    F: Fn(&I) -> SimpleInstr,
{
    let mut labelMap = BTreeMap::new();
    for (_, block) in &g.BasicBlocks {
        if let Some((_, first)) = block.instructions.first() {
            if let SimpleInstr::Label(label) = simplify(first) {
                labelMap.insert(label, block.id);
            }
        }
    }

    // Preserve the F# implementation's entry edge even for an empty input.
    addEdge(NodeId::Entry, NodeId::Block(0), g);
    let block_count = g.BasicBlocks.len();
    for index in 0..block_count {
        let (id_num, id, effect) = {
            let (id_num, block) = &g.BasicBlocks[index];
            let (_, last) = block
                .instructions
                .last()
                .expect("Internal error: empty basic block");
            (*id_num, block.id, simplify(last))
        };
        let next = if index + 1 == block_count {
            NodeId::Exit
        } else {
            NodeId::Block(id_num + 1)
        };
        match effect {
            SimpleInstr::Return => addEdge(id, NodeId::Exit, g),
            SimpleInstr::UnconditionalJump(target) => {
                let target_id = *labelMap.get(&target).expect("label not found in CFG");
                addEdge(id, target_id, g);
            }
            SimpleInstr::ConditionalJump(target) => {
                let target_id = *labelMap.get(&target).expect("label not found in CFG");
                addEdge(id, next, g);
                addEdge(id, target_id, g);
            }
            _ => addEdge(id, next, g),
        }
    }
}

pub fn instructionsToCfg<I, F>(
    simplify: F,
    debugLabel: impl Into<String>,
    instructions: Vec<I>,
) -> ControlFlowGraph<(), I>
where
    F: Fn(&I) -> SimpleInstr,
{
    let BasicBlocks = partitionIntoBasicBlocks(&simplify, instructions)
        .into_iter()
        .enumerate()
        .map(|(idx, instructions)| {
            (
                idx,
                BasicBlock {
                    id: NodeId::Block(idx),
                    instructions: instructions.into_iter().map(|i| ((), i)).collect(),
                    preds: Vec::new(),
                    succs: Vec::new(),
                    value: (),
                },
            )
        })
        .collect();
    let mut cfg = ControlFlowGraph {
        BasicBlocks,
        entrySuccs: Vec::new(),
        exitPreds: Vec::new(),
        debugLabel: debugLabel.into(),
    };
    addAllEdges(&simplify, &mut cfg);
    cfg
}

pub fn cfgToInstructions<V, I: Clone>(g: &ControlFlowGraph<V, I>) -> Vec<I> {
    g.BasicBlocks
        .iter()
        .flat_map(|(_, block)| block.instructions.iter().map(|(_, i)| i.clone()))
        .collect()
}

pub fn initializeAnnotation<V, I, W: Clone>(
    cfg: ControlFlowGraph<V, I>,
    dummyVal: W,
) -> ControlFlowGraph<W, I> {
    ControlFlowGraph {
        BasicBlocks: cfg
            .BasicBlocks
            .into_iter()
            .map(|(idx, block)| {
                (
                    idx,
                    BasicBlock {
                        id: block.id,
                        instructions: block
                            .instructions
                            .into_iter()
                            .map(|(_, i)| (dummyVal.clone(), i))
                            .collect(),
                        preds: block.preds,
                        succs: block.succs,
                        value: dummyVal.clone(),
                    },
                )
            })
            .collect(),
        entrySuccs: cfg.entrySuccs,
        exitPreds: cfg.exitPreds,
        debugLabel: cfg.debugLabel,
    }
}

pub fn stripAnnotations<V, I>(cfg: ControlFlowGraph<V, I>) -> ControlFlowGraph<(), I> {
    initializeAnnotation(cfg, ())
}

static COUNTER: AtomicUsize = AtomicUsize::new(0);

pub fn setCounter(counter: usize) {
    COUNTER.store(counter, Ordering::SeqCst);
}

pub fn getCounter() -> usize {
    COUNTER.load(Ordering::SeqCst)
}

fn writeNodeId(out: &mut dyn Write, id: NodeId) -> io::Result<()> {
    match id {
        NodeId::Exit => write!(out, "exit"),
        NodeId::Entry => write!(out, "entry"),
        NodeId::Block(n) => write!(out, "block{n}"),
    }
}

/// Write a Graphviz file and render it with `dot`, as the F# debug helper does.
pub fn printGraphviz<V, I, PI, PV>(
    ppInstr: PI,
    ppVal: PV,
    cfg: &ControlFlowGraph<V, I>,
) -> io::Result<()>
where
    PI: Fn(&mut dyn Write, &I) -> io::Result<()>,
    PV: Fn(&mut dyn Write, &V) -> io::Result<()>,
{
    let number = COUNTER.fetch_add(1, Ordering::SeqCst);
    let filename = format!("{}.{}.dot", cfg.debugLabel, number);
    let mut writer = File::create(&filename)?;
    writeln!(writer, "digraph {{")?;
    writeln!(writer, "  labeljust=l")?;
    writeln!(writer, "  node[shape=\"box\"]")?;
    writeln!(writer, "  entry[label=\"ENTRY\"]")?;
    writeln!(writer, "  exit[label=\"EXIT\"]")?;
    for (label, block) in &cfg.BasicBlocks {
        write!(writer, "block{label}[label=<<table><tr><td colspan=\"2\"><b>")?;
        writeNodeId(&mut writer, block.id)?;
        writeln!(writer, "</b></td></tr>")?;
        for (value, instruction) in &block.instructions {
            write!(writer, "<tr><td align=\"left\">")?;
            ppInstr(&mut writer, instruction)?;
            write!(writer, "</td><td align=\"left\">")?;
            ppVal(&mut writer, value)?;
            writeln!(writer, "</td></tr>")?;
        }
        write!(writer, "<tr><td colspan=\"2\">")?;
        ppVal(&mut writer, &block.value)?;
        writeln!(writer, "</td></tr></table>>]")?;
    }
    for successor in &cfg.entrySuccs {
        write!(writer, "entry -> ")?;
        writeNodeId(&mut writer, *successor)?;
        writeln!(writer)?;
    }
    for (label, block) in &cfg.BasicBlocks {
        for successor in &block.succs {
            write!(writer, "block{label} -> ")?;
            writeNodeId(&mut writer, *successor)?;
        }
        // The original emits one newline after all of a block's edges.
        writeln!(writer)?;
    }
    writeln!(writer, "}}")?;
    writer.flush()?;
    drop(writer);

    let png = format!("{}.{}.png", cfg.debugLabel, number);
    let status = Command::new("dot")
        .args(["-Tpng", &filename, "-o", &png])
        .status()?;
    if status.success() {
        Ok(())
    } else {
        Err(io::Error::other(format!(
            "graphviz fail: dot -Tpng {filename} -o {png}"
        )))
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[derive(Clone, Debug, Eq, PartialEq)]
    enum Instruction {
        Label(&'static str),
        Jump(&'static str),
        JumpIf(&'static str),
        Return,
        Other(u8),
    }

    fn simplify(i: &Instruction) -> SimpleInstr {
        match i {
            Instruction::Label(label) => SimpleInstr::Label((*label).into()),
            Instruction::Jump(label) => SimpleInstr::UnconditionalJump((*label).into()),
            Instruction::JumpIf(label) => SimpleInstr::ConditionalJump((*label).into()),
            Instruction::Return => SimpleInstr::Return,
            Instruction::Other(_) => SimpleInstr::Other,
        }
    }

    #[test]
    fn partitions_instructions_and_builds_edges() {
        let instructions = vec![
            Instruction::Label("start"),
            Instruction::Other(1),
            Instruction::JumpIf("done"),
            Instruction::Other(2),
            Instruction::Jump("start"),
            Instruction::Label("done"),
            Instruction::Return,
        ];
        let cfg = instructionsToCfg(simplify, "test", instructions.clone());

        assert_eq!(cfg.BasicBlocks.len(), 3);
        assert_eq!(cfg.entrySuccs, [NodeId::Block(0)]);
        assert_eq!(cfg.BasicBlocks[0].1.succs, [NodeId::Block(2), NodeId::Block(1)]);
        assert_eq!(cfg.BasicBlocks[1].1.succs, [NodeId::Block(0)]);
        assert_eq!(cfg.BasicBlocks[2].1.succs, [NodeId::Exit]);
        assert_eq!(cfg.exitPreds, [NodeId::Block(2)]);
        assert_eq!(cfgToInstructions(&cfg), instructions);
    }

    #[test]
    fn edge_updates_are_symmetric_and_do_not_duplicate_edges() {
        let mut cfg = instructionsToCfg(simplify, "test", vec![Instruction::Return]);
        addEdge(NodeId::Entry, NodeId::Block(0), &mut cfg);
        assert_eq!(cfg.entrySuccs, [NodeId::Block(0)]);
        assert_eq!(cfg.BasicBlocks[0].1.preds, [NodeId::Entry]);

        removeEdge(NodeId::Entry, NodeId::Block(0), &mut cfg);
        assert!(cfg.entrySuccs.is_empty());
        assert!(cfg.BasicBlocks[0].1.preds.is_empty());
    }

    #[test]
    fn annotations_can_change_type_and_be_stripped() {
        let cfg = instructionsToCfg(simplify, "test", vec![Instruction::Other(1)]);
        let annotated = initializeAnnotation(cfg, "unknown".to_owned());
        assert_eq!(annotated.BasicBlocks[0].1.value, "unknown");
        assert_eq!(annotated.BasicBlocks[0].1.instructions[0].0, "unknown");
        let stripped = stripAnnotations(annotated);
        assert_eq!(stripped.BasicBlocks[0].1.value, ());
    }
}
